apps/hash-api/src/integrations/linear/oauth.ts

```diff
@@ -0,0 +1,227 @@
+import { RequestHandler } from "express";
+
+import { createEntity } from "../../graph/knowledge/primitive/entity";
+import { isUserMemberOfOrg } from "../../graph/knowledge/system-types/user";
+import { SYSTEM_TYPES } from "../../graph/system-types";
+
+const linearClientId = process.env.LINEAR_CLIENT_ID;
+const linearClientSecret = process.env.LINEAR_CLIENT_SECRET;
+
+const linearOAuthCallbackUrl =
+  process.env.LINEAR_OAUTH_CALLBACK_URL ??
+  "http://localhost:5001/oauth/linear/callback";
+
+/**
+ * @todo oauth state will need to be store somewhere other than memory if we need:
+ *   - the authorization process to work if the server is rebooted/redeployed while the user is in the middle of it
+ *   - to be running multiple server instances without sticky sessions managed by the load balancer
+ *   - to be running multiple processes on the same server instance (e.g. Node Clusters) without ensuring
+ *        the same clients are handled by the same process
+ */
+const stateMap = new Map<
+  string,
+  {
+    actorEntityId: EntityId;
+    expires: Date;
+    ownedById: EntityUuid;
+    ownerType: "user" | "org";
+  }
+>();
+
+const generateLinearOAuthUrl = (oAuthState: string) => {
+  return (
+    "https://linear.app/oauth/authorize?" +
+    `client_id=${linearClientId}` +
+    `&redirect_uri=${linearOAuthCallbackUrl}` +
+    `&state=${oAuthState}` +
+    `&response_type=code&scope=write&prompt=consent`
+  );
+};
+
+export const oAuthLinear: RequestHandler<
+  {},
+  {},
+  {
+    ownedById?: EntityUuid;
+  }
+> =
+  // @todo upgrade to Express 5, which handles errors from async request handlers automatically
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  async (req, res) => {
+    if (!linearClientId) {
+      res
+        .status(501)
+        .send(
+          "Linear integration is not configured – set a client id and secret.",
+        );
+      return;
+    }
+
+    if (!req.user) {
+      res.status(401).send("You must be authenticated to do this.");
+      return;
+    }
+
+    const { ownedById } = req.query;
+
+    if (!ownedById) {
+      res.status(400).send("No ownedById for secret provided.");
+      return;
+    }
+
+    const userEntityId = req.user.entity.metadata.recordId.entityId;
+
+    if (
+      extractEntityUuidFromEntityId(userEntityId) !== ownedById &&
+      !(await isUserMemberOfOrg(req.context, {
+        userEntityId,
+        orgEntityUuid: ownedById as EntityUuid,
+      }))
+    ) {
+      res
+        .status(403)
+        .send(
+          "ownedById must represent the user or an organization they are a member of.",
+        );
+      return;
+    }
+
+    const ownerType =
+      extractEntityUuidFromEntityId(userEntityId) === ownedById
+        ? "user"
+        : "org";
+
+    const state = crypto.randomBytes(16).toString("hex");
+
+    stateMap.set(state, {
+      actorEntityId: req.user.entity.metadata.recordId.entityId,
+      expires: new Date(Date.now() + 1000 * 60 * 5),
```

Comment:
Can we add a comment how long this will be? Is it 5 minutes?

Reply:
Done in https://github.com/hashintel/hash/pull/2796/commits/c9d1bbc67ed7aac3bec813803ebe7d27b3bf110f